### PR TITLE
Implement `__setattr__` on `_FabricModule `

### LINKED ISSUE #17285: Implement `__setattr__` on `_FabricModule`
### Description & Motivation

We currently have only `__getattr__` implemented on the `_FabricModule`:

https://github.com/Lightning-AI/lightning/blob/c2d9ec4e1b9ab32b222630f3d094e657c65324f8/src/lightning/fabric/wrappers.py#L135-L143

Something like this works:

```py
model.x = 1
wrapped = fabric.setup(model)
print(wrapped.x)
```
But this doesn't:
```py
model.x = 1
wrapped = fabric.setup(model)
wrapped.x = 2
```
Current workaround is to do

```py
wrapped._original_module.x = 2
```

### Pitch

Implement  `__setattr__` on the `_FabricModule` and forward to the `_original_module`. This has to done carefully, as not to break the initialization of the `_FabricModule` and its attributes.

### Alternatives

_No response_

### Additional context

_No response_

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by ishandutta0098**
## What does this PR do?

I have implemented `__setattr__` for `_FabricModule` 

Fixes #17285 

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by ishandutta0098** [comment]
Review Request: @awaelchli @justusschock

File: tests/tests_fabric/test_wrappers.py [hunk 1]
```diff
@@ -353,3 +353,36 @@ def test_is_wrapped():
...
+    """Test that setattr sets attributes on the original module."""
+
+    class OriginalModule(torch.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.layer = torch.nn.Linear(2, 3)
+            self.attribute = 1
+
+        def method(self):
+            return 2
```
Comment (awaelchli) [inline tests/tests_fabric/test_wrappers.py:368]:
```suggestion
```
I don't see it being used anywhere

File: src/lightning/fabric/wrappers.py [hunk 2]
```diff
@@ -142,6 +142,16 @@ def __getattr__(self, item: Any) -> Any:
             original_module = super().__getattr__("_original_module")
             return getattr(original_module, item)
 
+    def __setattr__(self, name: str, value: Any) -> None:
+        # Check if _FabricModule is fully initialized
+        if "_original_module" in self.__dict__:
+            original_module = super().__getattr__("_original_module")
```
Comment (awaelchli) [inline src/lightning/fabric/wrappers.py:148]: Would this work too?
```suggestion
        if hasattr(self, "_original_module"):
            original_module = self._original_module
```

File: tests/tests_fabric/test_wrappers.py [hunk 3]
```diff
@@ -353,3 +353,36 @@ def test_is_wrapped():
     assert not is_wrapped(dataloader)
     wrapped = _FabricDataLoader(dataloader)
     assert is_wrapped(wrapped)
+
+
+def test_fabric_module_setattr():
```
Comment (awaelchli) [inline tests/tests_fabric/test_wrappers.py:358]: Could you place this test below the test named `test_fabric_module_attribute_lookup`? It would be good to keep them close :)

**Review by awaelchli — COMMENTED** [review]
This looks great!

There are two some failing tests. We need to figure out the failing case related to nn.Module's override of setattr.

File: tests/tests_fabric/test_wrappers.py (hunk 1 above)
Comment (ishandutta0098) [reply tests/tests_fabric/test_wrappers.py]: I have removed this and committed

File: src/lightning/fabric/wrappers.py (hunk 2 above)
Comment (ishandutta0098) [reply src/lightning/fabric/wrappers.py]: I have updated this and committed

File: tests/tests_fabric/test_wrappers.py (hunk 3 above)
Comment (ishandutta0098) [reply tests/tests_fabric/test_wrappers.py]: I have changed the location and committed

**Comment by ishandutta0098** [comment]
@awaelchli I have solved the failing case due to the overrides in `__setattr__`, the test case passes for it now. 
However I am getting an issue at a different test case `test_fabric_module_device_dtype_propagation` in `test_wrappers.py`. 
It seems to me that the `dtype` is not getting propagated properly. 

I tried adding special cases in the `__setattr__` as well as overriding the `to` function within the `_FabricModule` but the error persists (I didn't commit this), do you have any inputs for this?

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Comment by bkiat1123** [comment]
Saw this PR while trying to debug Fabric issue.

To recap things to taken care of when implementing `__setattr__` method

- It should not break the initialization
- Instance should still able to change own attribute after initialization
- We should able to set attribute on original_module by `wrapped.name = value`

First, let’s get initialization out of the way. We can do what PyTorch does in initialization of `nn.Module` , using super class to initialize the attribute. This way we don’t have to worry about `__setattr__` messing with initialization.

```python
class _FabricModule(_DeviceDtypeModuleMixin):
    def __init__(self, ...):
        super().__init__()
        super().__setattr__("_forward_module", forward_module)
        super().__setattr__("_original_module", original_module or forward_module)
        super().__setattr__("_precision", precision)
```

The `super()` will use `__setattr__` method from `nn.Module`, which themselves avoid using their own `__setattr` to avoid overhead. We could use `object.__setattr__(self, "name", "value")` syntax to avoid the overhead. Personally, I would prefer the `super()` for readability.

Note that the code in this PR failed to set precision with the unmodified `__setattr__`.

The next thing is resolution order. 

Both `_DeviceDtypeModuleMixin` and `original_module` inherits from `nn.Module`, so they both share a set of attributes that originated from `nn.Module`. Whenever `setattr` is called on attribute names that exist in `nn.Module`, we set it on original_module instead of the wrapper.

Other than that, we assume no attribute name clash between `_FabricModule` and `original_module`, otherwise we are in deep water. Another assumption is we don’t need to add new attribute on wrapper, only modifying existing attribute.

Another gotcha to watch out is when we check attribute of the wrapper using `hasattr`. It calls `__getattr__` under the hood, which we have modified to check original module too.

The resolution order we can use are

- If attribute name is in original module, including nn.Module attributes, set it in original module
- If it is in wrapper, set it in the wrapper.
- If neither has the attribute, set it in original module

The code will look like
```python
def __setattr__(self, name: str, value: Any) -> None:
    # Get the _original_module attribute
    original_module = getattr(self, "_original_module")

    if hasattr(original_module, name):
        setattr(original_module, name, value)

    # Since the above is False, we are only checking the wrapper here
    elif hasattr(self, name)
        super().__setattr__(name, value)
    else:
        setattr(original_module, name, value)
```

**Comment by awaelchli** [comment]
@bkiat1123  Thanks for looking into this. You could test out your changes by sending a PR and letting all tests run against the change, or it could be added to this PR @ishandutta0098. In any case, a contribution for this would be welcomed.

**Comment by awaelchli** [comment]
This was finished in #17285
Thanks @bkiat1123 and @ishandutta0098
